[stdlib] availability annotations for popFirst; drop Unicode 8 tries (finally)

milseman opened the pull request:
<!-- What's in this pull request? -->
This adds availability diagnostics for String.*View.popFirst, which is mostly unavailable these days. It also drops the Unicode 8 tries, switching all code paths to ICU for grapheme breaking.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

milseman committed:
[stdlib] Drop Unicode 8 tries in stdlib (entirely)

This drops the last vestage of Unicode 8 tries from the standard
library. Switches everything over to use ICU.

milseman committed:
[stdlib] Availability checking for String.*View.popFirst

Collection's popFirst is only present when the Collection is its own
SubSequence type. String and String's views are no longer their own
SubSequenes, so popFirst is no longer present. Unfortunately, this
breaks code in swift-version 3 and it also gives a terrible diagnostic
to users. This change introduces an implementation for swift-version 3
mode and better diagnostics for Swift 4 code. Tests included.

milseman commented:
@swift-ci please test

milseman commented:
Ugh, nasty fragile associated types bites again. We can't enforce the unavailability of `popFirst` on String.CharacterView while keeping String.CharacterView its own SubSequence. Will probably just remove that check for now and keep it available on String.CharacterView until String.CharacterView is properly deprecated.

milseman commented:
@swift-ci please test

moiseev commented on stdlib/public/core/Substring.swift.gyb:
```diff
@@ -612,3 +612,39 @@ extension Substring {
+// popFirst() is only present when a collection is its own subsequence. This was
+// dropped in Swift 4.
+extension String {
+  @available(swift, obsoleted: 4, message:
```
As discussed, it probably makes sense to add a `deprecated: 3.2` here as well. It will still compile and work in Swift 3.2 as is, but the Swift 4 compatible alternative is also available in Swift 3.2, so why not let developers know that and thus simplify the transition to Swift 4 eventually.

moiseev commented on test/stdlib/StringCompatibility.swift:
```diff
@@ -356,5 +356,20 @@ Tests.test("LosslessStringConvertible/force unwrap/\(swift)") {
+#if !swift(>=4)
+Tests.test("popFirst") {
+  var str = "abcdef"
+  expectEqual("a", str.popFirst()!)
```
FYI: there is `expectOptionalEqual` family of functions in `StdlibUnittest`. Not big deal, but they will fail properly instead of reporting an unsafe unwrap in case something goes wrong.

moiseev commented on test/stdlib/StringCompatibility.swift:
```diff
@@ -356,5 +356,20 @@ Tests.test("LosslessStringConvertible/force unwrap/\(swift)") {
+  expectEqual("d", str.popFirst()!)
+  expectEqual("e", str.popFirst()!)
+  expectEqual("f", str.popFirst()!)
+  expectEqual(nil, str.popFirst())
```
There is also `expectNil`. `StdlibUnittest` is actually quite rich =)

moiseev reviewed (changes requested):
I think `deprecated:` is the right thing to do, other comments are minor. LGTM otherwise!

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please test OS X Platform

milseman commented:
@moiseev, does this look good now? If so, could you un-block the review so we can merge if testing passes?

moiseev commented on stdlib/public/core/Substring.swift.gyb:
```diff
@@ -612,3 +612,42 @@ extension Substring {
+// popFirst() is only present when a collection is its own subsequence. This was
+// dropped in Swift 4.
+extension String {
+  @available(swift, deprecated: 3.2)
```
The problem is that this useful message will not be displayed as part of deprecation warning.
It is possible to combine `deprecated` and `obsoleted` in a single attribute though, this way the message will be shared.

moiseev commented:
@milseman Sorry, I missed the moment when you pushed the changes. Would have commented earlier.

milseman commented:
@swift-ci please test OS X Platform

milseman committed:
[stdlib] Deprecate popFirst in Swift 3.2

Add a deprecation warning for use of popFirst even with -swift-version
3, as it's prone to leaks under the old model. Test cases added.

milseman commented:
@moiseev ping :-)

milseman commented:
@swift-ci please test

shahmishal commented:
@swift-ci please test

moiseev reviewed (approved):
:+1:
